<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.538">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Lun M Wong">
<meta name="author" content="Qi-Yong H Ai">
<meta name="author" content="Tiffany Y So">
<meta name="author" content="Jacky WK Lam">
<meta name="author" content="Ann D King">
<meta name="dcterms.date" content="2023-06-06">

<title>Computer vision, biostatistics, and AI in Radiology - Explainable artificial intelligence for automatic detection of early nasopharyngeal carcinoma on MRI</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<link href="../../../favicon.png" rel="icon" type="image/png">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="dark">
<link href="../../../site_libs/bootstrap/bootstrap-dark.min.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      // ...your configuration options...
      "HTML-CSS": { fonts: ["TeX"] }
    });
</script>
    
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&amp;display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Georgia:wght@400;700&amp;display=swap" rel="stylesheet">


<link rel="stylesheet" href="../../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">Computer vision, biostatistics, and AI in Radiology</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../about.html"> 
<span class="menu-text">About Me</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ml-w"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://scholar.google.com/citations?user=JNKeB6cAAAAJ&amp;hl=en"> <i class="bi bi-mortarboard" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="http://www.diir.cuhk.edu.hk/profile/mr-wong-lun-matthew"> <i class="bi bi-person-circle" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    <div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="01-xai-npc-bn.pdf"><i class="bi bi-file-pdf"></i>Typst</a></li></ul></div></div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Posts</li><li class="breadcrumb-item"><a href="../../../posts/deep-learning/explainable-ai/index.html">Deep Learning</a></li><li class="breadcrumb-item"><a href="../../../posts/deep-learning/explainable-ai/index.html">Explainable AI (XAI)</a></li><li class="breadcrumb-item"><a href="../../../posts/deep-learning/explainable-ai/01-xai-npc-bn.html">Explainable artificial intelligence for automatic detection of early nasopharyngeal carcinoma on MRI</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Explainable artificial intelligence for automatic detection of early nasopharyngeal carcinoma on MRI</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Lun M Wong <a href="mailto:lun.m.wong+gio@cuhk.edu.hk" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-6047-4609" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            The Chinese University of Hong Kong
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Qi-Yong H Ai </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            The Chinese University of Hong Kong
          </p>
        <p class="affiliation">
            The Hong Kong Polytechnic University
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Tiffany Y So </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            The Chinese University of Hong Kong
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Jacky WK Lam </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            The Chinese University of Hong Kong
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Ann D King </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            The Chinese University of Hong Kong
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">June 6, 2023</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="block-title">Synopsis</div>
    In this study, we propose a simple method to improve the explainability of artificial intelligence, specifically convolutional neural networks (CNNs), for the automatic detection of early nasopharyngeal carcinoma (NPC) on magnetic resonance imaging (MRI). We show a long-short-term-memory (LSTM) unit can be introduced into a CNN to read 3-dimensional medical image series. A risk curve can be extracted from the LSTM to visualize the “thought process” of the network when it reads through the input MRI slice-by-slice. This modification improves the explainability of the network without reducing performance for the early NPC detections of the original CNN.
  </div>
</div>


</header>


<section id="synopsis" class="level1">
<h1>Synopsis</h1>
<p>In this study, we propose a simple method to improve the explainability of artificial intelligence, specifically convolutional neural networks (CNNs), for the automatic detection of early nasopharyngeal carcinoma (NPC) on magnetic resonance imaging (MRI). We show a long-short-term-memory (LSTM) unit can be introduced into a CNN to read 3-dimensional medical image series. A risk curve can be extracted from the LSTM to visualize the “thought process” of the network when it reads through the input MRI slice-by-slice. This modification improves the explainability of the network without reducing performance for the early NPC detections of the original CNN.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>This work was presented during ISMRM 2023, in Toronto, CA.</p>
</div>
</div>
</section>
<section id="abstract" class="level1">
<h1>Abstract</h1>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Early detection of nasopharyngeal carcinoma (NPC) can markedly reduce the mortality and morbidity of treatment complications. Magnetic resonance imaging (MRI) can detect up to 38.1% more early NPCs when compared to the endoscopy <span class="citation" data-cites="RN438 RN131">(<a href="#ref-RN438" role="doc-biblioref">King et al. 2020</a>; <a href="#ref-RN131" role="doc-biblioref">Liu et al. 2021</a>)</span>, which is the current investigation of choice for individuals who are Epstein-Barr Virus DNA positive on an NPC screening blood test <span class="citation" data-cites="RN217 RN446">(<a href="#ref-RN217" role="doc-biblioref">K. C. A. Chan et al. 2017</a>; <a href="#ref-RN446" role="doc-biblioref">D. C. T. Chan et al. 2022</a>)</span>. To expand the role of MRI in early detection, our team has previously investigated artificial intelligence (AI) algorithms to automatically diagnose and reduce costs <span class="citation" data-cites="RN397 RN394 RN437">(<a href="#ref-RN397" role="doc-biblioref">Wong, Ai, Mo, et al. 2021</a>; <a href="#ref-RN394" role="doc-biblioref">Wong, Ai, Poon, et al. 2021</a>; <a href="#ref-RN437" role="doc-biblioref">Wong, King, et al. 2021</a>)</span>. However, despite these algorithms as well as those proposed by other research teams <span class="citation" data-cites="RN295 RN396">(<a href="#ref-RN295" role="doc-biblioref">Ke et al. 2020</a>; <a href="#ref-RN396" role="doc-biblioref">Deng et al. 2022</a>)</span> performed remarkably in these tasks, they inevitably lack explainability because of their complex model structure comprising millions of learnable parameters. The explainability issue is currently one of the major factors that has prevented the implementation of these deep learning methods in clinical workflows.</p>
<p>In this study, we propose a method to improve the explainability of convolutional neural network (CNN) algorithms for the automatic diagnosis of NPC on MRI. It requires simply an additional long short-term memory (LSTM), which enables visualization of the network’s “thought process” and localization of key slices that determined the AI’s prediction and could aid understanding of the results.</p>
</section>
<section id="methods" class="level2">
<h2 class="anchored" data-anchor-id="methods">Methods</h2>
<section id="model-architecture" class="level3">
<h3 class="anchored" data-anchor-id="model-architecture">Model architecture</h3>
<p>We introduced an LSTM unit to our previously proposed slice-wise residual attention network (SWRAN) <span class="citation" data-cites="RN437 RN113">(<a href="#ref-RN437" role="doc-biblioref">Wong, King, et al. 2021</a>; <a href="#ref-RN113" role="doc-biblioref">Wang et al. 2017</a>)</span>, inserted before the output linear layer and replacing the original max-pooling layer for improving explainability. On receiving MRI images, the original SWRAN encoded the input slice-by-slice which were then read sequentially by the additional LSTM to identify NPC. The prediction obtained from individual slices has the advantage over a single prediction from all slices in that it can indicate how the network shifted between an NPC positive or negative prediction while it scans through the slices. This shift can be visualized as a “risk curve” to improve explainability. Detailed architecture and a legend of the risk curve is given in <a href="#fig-cnnrnn" class="quarto-xref">Figure&nbsp;1</a> and <a href="#fig-lstm" class="quarto-xref">Figure&nbsp;2</a>, respectively. This network was trained in two steps. First, the original SWRAN without the LSTM was pre-trained. Then, the convolutional kernels were fixed to finetune only the LSTM and the output layer of the network (<a href="#fig-cnnrnn" class="quarto-xref">Figure&nbsp;1</a>).</p>
<div id="fig-cnnrnn" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-cnnrnn-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/ISMRM2023-002385_Fig1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-cnnrnn-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Network architecture of the proposed method for detecting nasopharyngeal carcinoma (NPC) on MRI. The original slice-wise residual attention network (SWRAN) encodes each slice of the input into deep feature vectors. During pre-train, the original output layer of SWRAN is invoked. During finetuning, the encoded vectors are sequentially fed into the long short-term memory (LSTM) unit. Its output after reading all slices is the final prediction, whereas the individual hidden states is the risk curve that visualize the “thought process” of the network.
</figcaption>
</figure>
</div>
</section>
<section id="patients-for-validation-of-performance" class="level3">
<h3 class="anchored" data-anchor-id="patients-for-validation-of-performance">Patients for validation of performance</h3>
<p>We retrospectively included 884 patients previously scanned for suspected NPC using the T2-weighted fat-suppressed sequence in our center on either a 1.5T or 3.0T MRI scanner, of which 316 had stage I/II NPC and 568 had benign hyperplasia/normal nasopharynx without NPC. These patients were divided with stratification to NPC status into training, testing and validation sets at a ratio of roughly 7:2:1 (620:176:88). Finally, the diagnostic performance of the pre-trained and finetuned network was evaluated over the testing set through receiver operator characteristics (ROC) analysis.</p>
<div id="fig-lstm" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-lstm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/ISMRM2023-002385_Fig2.png" class="img-fluid figure-img" style="width:80.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-lstm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Legend of the risk curve plotted using the hidden states of the long short-term memory (LSTM). There are three elements in this plot: (a) the risk curve, (b) the reference threshold, and (c) a vertical line marking the current slice. As the LSTM scans through the slices, it can respond to malignant features and predict higher risks of NPC, which stays high for subsequent slices that without tumor, showing the LSTM registers malignant features seen in its memory. The final prediction is taken at the end of the curve at the right, area above the threshold is the NPC +ve zone (red shade).
</figcaption>
</figure>
</div>
</section>
<section id="assessment-of-the-risk-curves" class="level3">
<h3 class="anchored" data-anchor-id="assessment-of-the-risk-curves">Assessment of the risk curves</h3>
<p>An expert in head and neck imaging inspected the risk curves of the testing set and summarized qualitatively the patterns and characteristics of the curves in NPC and benign hyperplasia/normal subjects.</p>
</section>
</section>
<section id="results" class="level2">
<h2 class="anchored" data-anchor-id="results">Results</h2>
<section id="diagnostic-performance" class="level3">
<h3 class="anchored" data-anchor-id="diagnostic-performance">Diagnostic performance</h3>
<p>Sensitivity, specificity, accuracy and area under the ROC curve (AUC) of the pre-trained network were 91.3%, 94.4%, 93.2% and 0.981, respectively; and those of the fine-tuned network with LSTM were 95.7%, 91.6%, 93.2% and 0.976, respectively.</p>
</section>
<section id="assessment-of-the-risk-curve" class="level3">
<h3 class="anchored" data-anchor-id="assessment-of-the-risk-curve">Assessment of the risk curve</h3>
<p>In 63 out of 66 true-positive predictions of NPC cases, the risk curves were one of two forms: (i) started off flat but sharp increase when reaching the primary NPC (n=29); (ii) raised slowly from the beginning till the end (n=34). In 68 out of 98 true-negative predictions, the risk curves stayed flat under the reference line unremarkably, the rest fluctuates but eventually settled below the referenced threshold after all slices were read. In 2 out of 3 false-negative cases, the risk curves showed a local peak at the primary NP which might allow manual rectification. Representative examples are provided in <a href="#fig-npc-bn-compare" class="quarto-xref">Figure&nbsp;3</a>. The slope and value of the risk curves were exploited to locate and highlight key slices that lead to an NPC positive prediction shown in <a href="#fig-results1" class="quarto-xref">Figure&nbsp;4</a>. Examples of interpreting the risk curves are given in <a href="#fig-results2" class="quarto-xref">Figure&nbsp;5</a>.</p>
<div id="fig-npc-bn-compare" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-npc-bn-compare-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/ISMRM2023-002385_Fig3.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-npc-bn-compare-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Representative examples of the risk curves plotted on the axial MRI of two patients with nasopharyngeal carcinoma (NPC) (left) and with benign hyperplasia (right). The x- and y-axis of the risk curves are the slice index and predict risk respectively. The green vertical line marks the position of the current slice on the risk plot. As shown on the left, the network responded and the risk value raised when it detected slices with NPC (white arrow), whereas the curve typically stays under the reference threshold for the benign cases like the one on the right.
</figcaption>
</figure>
</div>
<div id="fig-results1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-results1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/ISMRM2023-002385_Fig4.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-results1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Auxiliary visual effects designed to highlight the slices that triggers the nasopharyngeal carcinoma (NPC) positive prediction by automatically interpreting the slope of the risk curve. Slices are order from left to right and top to bottom. The slope of the risk curve is used to locate critical slices that lead to the network’s prediction. Here we empirically set a threshold for slope to automatically highlight slices as amber, suggesting the network identify suspicious malignant features, and a higher level of slope to highlight slices as red, suggesting the network.
</figcaption>
</figure>
</div>
<div id="fig-results2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-results2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/ISMRM2023-002385_Fig5.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-results2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: More examples of interpretating the risk curves of two nasopharyngeal carcinoma (NPC) and a benign hyperplasia patient in the testing set. (a) true-positive prediction for NPC, the network responded to the enlarged left retropharyngeal node (blue arrow) that is near the tumor (green arrows) by rising risk values across slices; (b) false-negative prediction, but the risk curve reflected suspicious thickening (blue arrows) briefly; (c) a true-negative prediction, the benign hyperplasia was recognized and lighted up a slice, but the risk curve is otherwise unremarkable.
</figcaption>
</figure>
</div>
</section>
</section>
<section id="discussion" class="level2">
<h2 class="anchored" data-anchor-id="discussion">Discussion</h2>
<p>We have proposed a simple method using LSTM to improve the explainability of CNN for early NPC detection on MRI in this study that achieved an AUC of 0.975, which is similar to the CNN without the LSTM (AUC=0.981). Compared to other methods employed to explain classification predictions by CNN in the literature, such as the attention mechanism and guided back-propagation <span class="citation" data-cites="RN444">(<a href="#ref-RN444" role="doc-biblioref">Springenberg et al. 2014</a>)</span> and Grad-CAM <span class="citation" data-cites="RN442">(<a href="#ref-RN442" role="doc-biblioref">Selvaraju et al., n.d.</a>)</span>, our proposed method has the advantage of removing the arbitrariness from the need to pick a layer for visualization. It can add also insight into cases that produce false-positive or negative results which may enable modifications to future networks to improve performance. This study has some limitations. Firstly, this study did not address quantitatively the improvements in explainability owing to the lack of well-established metrics or systematic methods to do so. Secondly, the addition of LSTM increases training time and difficulty because the network needs to be trained in two steps.</p>
</section>
<section id="conclusion" class="level2">
<h2 class="anchored" data-anchor-id="conclusion">Conclusion</h2>
<p>LSTM introduced to CNN can improve the explainability without compromising performance in automatic early NPC detection on MRI. Acknowledgements We would like to acknowledge Mr.&nbsp;Yip Man Tsang for his effort to collect the imaging dataset that is used in this study.</p>



</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-RN446" class="csl-entry" role="listitem">
Chan, D. C. T., W. K. J. Lam, E. P. Hui, B. B. Y. Ma, C. M. L. Chan, V. C. T. Lee, S. H. Cheng, et al. 2022. <span>“Improved Risk Stratification of Nasopharyngeal Cancer by Targeted Sequencing of Epstein-Barr Virus DNA in Post-Treatment Plasma.”</span> Journal Article. <em>Ann Oncol</em> 33 (8): 794–803. <a href="https://doi.org/10.1016/j.annonc.2022.04.068">https://doi.org/10.1016/j.annonc.2022.04.068</a>.
</div>
<div id="ref-RN217" class="csl-entry" role="listitem">
Chan, K. C. A., J. K. S. Woo, A. King, B. C. Y. Zee, W. K. J. Lam, S. L. Chan, S. W. I. Chu, et al. 2017. <span>“Analysis of Plasma Epstein-Barr Virus DNA to Screen for Nasopharyngeal Cancer.”</span> Journal Article. <em>N Engl J Med</em> 377 (6): 513–22. <a href="https://doi.org/10.1056/NEJMoa1701717">https://doi.org/10.1056/NEJMoa1701717</a>.
</div>
<div id="ref-RN396" class="csl-entry" role="listitem">
Deng, Y. S., C. F. Li, X. Lv, W. X. Xia, L. J. Shen, B. Z. Jing, B. Li, et al. 2022. <span>“The Contrast-Enhanced MRI Can Be Substituted by Unenhanced MRI in Identifying and Automatically Segmenting Primary Nasopharyngeal Carcinoma with the Aid of Deep Learning Models: An Exploratory Study in Large-Scale Population of Endemic Area.”</span> Journal Article. <em>Computer Methods and Programs in Biomedicine</em> 217: 106702. https://doi.org/<a href="https://doi.org/ARTN 106702
10.1016/j.cmpb.2022.106702">ARTN 106702 10.1016/j.cmpb.2022.106702</a>.
</div>
<div id="ref-RN295" class="csl-entry" role="listitem">
Ke, L., Y. Deng, W. Xia, M. Qiang, X. Chen, K. Liu, B. Jing, et al. 2020. <span>“Development of a Self-Constrained 3D DenseNet Model in Automatic Detection and Segmentation of Nasopharyngeal Carcinoma Using Magnetic Resonance Images.”</span> Journal Article. <em>Oral Oncol</em> 110: 104862. <a href="https://doi.org/10.1016/j.oraloncology.2020.104862">https://doi.org/10.1016/j.oraloncology.2020.104862</a>.
</div>
<div id="ref-RN438" class="csl-entry" role="listitem">
King, A. D., J. K. S. Woo, Q. Y. Ai, F. K. F. Mo, T. Y. So, W. K. J. Lam, I. O. L. Tse, et al. 2020. <span>“Early Detection of Cancer: Evaluation of MR Imaging Grading Systems in Patients with Suspected Nasopharyngeal Carcinoma.”</span> Journal Article. <em>AJNR Am J Neuroradiol</em> 41 (3): 515–21. <a href="https://doi.org/10.3174/ajnr.A6444">https://doi.org/10.3174/ajnr.A6444</a>.
</div>
<div id="ref-RN131" class="csl-entry" role="listitem">
Liu, Z., H. Li, K. J. Yu, S. H. Xie, A. D. King, Q. H. Ai, W. J. Chen, et al. 2021. <span>“Comparison of New Magnetic Resonance Imaging Grading System with Conventional Endoscopy for the Early Detection of Nasopharyngeal Carcinoma.”</span> Journal Article. <em>Cancer</em> 127 (18): 3403–12. <a href="https://doi.org/10.1002/cncr.33552">https://doi.org/10.1002/cncr.33552</a>.
</div>
<div id="ref-RN442" class="csl-entry" role="listitem">
Selvaraju, Ramprasaath R, Michael Cogswell, Abhishek Das, Ramakrishna Vedantam, Devi Parikh, and Dhruv Batra. n.d. <span>“Grad-Cam: Visual Explanations from Deep Networks via Gradient-Based Localization.”</span> Conference Proceedings. In <em>Proceedings of the IEEE International Conference on Computer Vision</em>, 618–26.
</div>
<div id="ref-RN444" class="csl-entry" role="listitem">
Springenberg, Jost Tobias, Alexey Dosovitskiy, Thomas Brox, and Martin Riedmiller. 2014. <span>“Striving for Simplicity: The All Convolutional Net.”</span> Web Page. <a href="https://doi.org/10.48550/arXiv.1412.6806">https://doi.org/10.48550/arXiv.1412.6806</a>.
</div>
<div id="ref-RN113" class="csl-entry" role="listitem">
Wang, Fei, Mengqing Jiang, Chen Qian, Shuo Yang, Cheng Li, Honggang Zhang, Xiaogang Wang, and Xiaoou Tang. 2017. <span>“Residual Attention Network for Image Classification.”</span> Online Multimedia.
</div>
<div id="ref-RN397" class="csl-entry" role="listitem">
Wong, L. M., Q. Y. H. Ai, F. K. F. Mo, D. M. C. Poon, and A. D. King. 2021. <span>“Convolutional Neural Network in Nasopharyngeal Carcinoma: How Good Is Automatic Delineation for Primary Tumor on a Non-Contrast-Enhanced Fat-Suppressed T2-Weighted MRI?”</span> Journal Article. <em>Jpn J Radiol</em> 39 (6): 571–79. <a href="https://doi.org/10.1007/s11604-021-01092-x">https://doi.org/10.1007/s11604-021-01092-x</a>.
</div>
<div id="ref-RN394" class="csl-entry" role="listitem">
Wong, L. M., Q. Y. H. Ai, D. M. C. Poon, M. Tong, B. B. Y. Ma, E. P. Hui, L. Shi, and A. D. King. 2021. <span>“A Convolutional Neural Network Combined with Positional and Textural Attention for the Fully Automatic Delineation of Primary Nasopharyngeal Carcinoma on Non-Contrast-Enhanced MRI.”</span> Journal Article. <em>Quant Imaging Med Surg</em> 11 (9): 3932–44. <a href="https://doi.org/10.21037/qims-21-196">https://doi.org/10.21037/qims-21-196</a>.
</div>
<div id="ref-RN437" class="csl-entry" role="listitem">
Wong, L. M., A. D. King, Q. Y. H. Ai, W. K. J. Lam, D. M. C. Poon, B. B. Y. Ma, K. C. A. Chan, and F. K. F. Mo. 2021. <span>“Convolutional Neural Network for Discriminating Nasopharyngeal Carcinoma and Benign Hyperplasia on MRI.”</span> Journal Article. <em>Eur Radiol</em> 31 (6): 3856–63. <a href="https://doi.org/10.1007/s00330-020-07451-y">https://doi.org/10.1007/s00330-020-07451-y</a>.
</div>
</div></section><section class="quarto-appendix-contents" id="quarto-citation"><h2 class="anchored quarto-appendix-heading">Citation</h2><div><div class="quarto-appendix-secondary-label">BibTeX citation:</div><pre class="sourceCode code-with-copy quarto-appendix-bibtex"><code class="sourceCode bibtex">@inproceedings{m. l. and ai, q. y. h. and so, t. y. and lam, w. k. j. and king, a. d.2023,
  author = {M. L. and Ai, Q. Y. H. and So, T. Y. and Lam, W. K. J. and
    King, A. D. , Wong},
  title = {Explainable Artificial Intelligence for Automatic Detection
    of Early Nasopharyngeal Carcinoma on {MRI}},
  booktitle = {Proceedings of the International Society for Magnetic
    Resonance in Medicine},
  volume = {31},
  date = {2023-06-06},
  langid = {en},
  abstract = {In this study, we propose a simple method to improve the
    explainability of artificial intelligence, specifically
    convolutional neural networks (CNNs), for the automatic detection of
    early nasopharyngeal carcinoma (NPC) on magnetic resonance imaging
    (MRI). We show a long-short-term-memory (LSTM) unit can be
    introduced into a CNN to read 3-dimensional medical image series. A
    risk curve can be extracted from the LSTM to visualize the “thought
    process” of the network when it reads through the input MRI
    slice-by-slice. This modification improves the explainability of the
    network without reducing performance for the early NPC detections of
    the original CNN.}
}
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre><div class="quarto-appendix-secondary-label">For attribution, please cite this work as:</div><div id="ref-m. l. and ai, q. y. h. and so, t. y. and lam, w. k. j. and king, a. d.2023" class="csl-entry quarto-appendix-citeas" role="listitem">
M. L. and Ai, Q. Y. H. and So, T. Y. and Lam, W. K. J. and King, A. D.,
Wong. 2023. <span>“Explainable Artificial Intelligence for Automatic
Detection of Early Nasopharyngeal Carcinoma on MRI.”</span> In
<em>Proceedings of the International Society for Magnetic Resonance in
Medicine</em>. Vol. 31.
</div></div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = true;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>©Copyright 2024, MLun Wong</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>